code/datums/helper_datums/teleport.dm

```diff
@@ -202,3 +193,19 @@
 	var/list/turfs = get_teleport_turfs(center, precision)
 	if (length(turfs))
 		return pick(turfs)
+
+/// Validates that the teleport being attempted is valid or not
+/proc/check_teleport_valid(atom/tele_atom, atom/destination, area/from_area, area/to_area, turf/from_turf, turf/to_turf, channel)
+	if(HAS_TRAIT(tele_atom, TRAIT_NO_TELEPORT))
+		return FALSE
+
+	if((from_area.area_flags & NOTELEPORT) || (to_area.area_flags & NOTELEPORT))
+		return FALSE
+
+	if(SEND_SIGNAL(tele_atom, COMSIG_MOVABLE_TELEPORTED, destination, channel) & COMPONENT_BLOCK_TELEPORT)
+		return FALSE
+
+	if(SEND_SIGNAL(to_turf, COMSIG_ATOM_INTERCEPT_TELEPORT, channel, from_turf, to_turf) & COMPONENT_BLOCK_TELEPORT)
+		return FALSE
```

Comment:
You didn't write this code and fixing it is a hassle, but this is terrible and broken because things can register to these signals expecting them to imply a completed teleport

Reply:
Fixing